llama-index-integrations/readers/llama-index-readers-azstorage-blob/llama_index/readers/azstorage_blob/base.py

```diff
@@ -64,102 +71,167 @@ class AzStorageBlobReader(BasePydanticReader):
     def class_name(cls) -> str:
         return "AzStorageBlobReader"
 
-    def load_data(self) -> List[Document]:
-        """Load file(s) from Azure Storage Blob."""
+    def _get_container_client(self):
         if self.connection_string:
-            container_client = ContainerClient.from_connection_string(
+            return ContainerClient.from_connection_string(
                 conn_str=self.connection_string,
                 container_name=self.container_name,
             )
-        else:
-            container_client = ContainerClient(
-                self.account_url, self.container_name, credential=self.credential
-            )
-        total_download_start_time = time.time()
+        return ContainerClient(
+            self.account_url, self.container_name, credential=self.credential
+        )
+
+    def _download_files_and_extract_metadata(self, temp_dir: str) -> Dict[str, Any]:
+        """Download files from Azure Storage Blob and extract metadata."""
+        container_client = self._get_container_client()
         blob_meta = {}
 
-        with tempfile.TemporaryDirectory() as temp_dir:
-            if self.blob:
-                blob_client = container_client.get_blob_client(self.blob)
-                stream = blob_client.download_blob()
-                sanitized_file_name = stream.name.replace("/", "-")
+        if self.blob:
+            blob_client = container_client.get_blob_client(self.blob)
+            stream = blob_client.download_blob()
+            sanitized_file_name = stream.name.replace("/", "-")
+            download_file_path = os.path.join(temp_dir, sanitized_file_name)
+            logger.info(f"Start download of {self.blob}")
+            start_time = time.time()
+            with open(file=download_file_path, mode="wb") as download_file:
+                stream.readinto(download_file)
+            blob_meta[sanitized_file_name] = blob_client.get_blob_properties()
+            end_time = time.time()
+            logger.info(
+                f"{self.blob} downloaded in {end_time - start_time} seconds."
+            )
+        else:
+            logger.info("Listing blobs")
+            blobs_list = container_client.list_blobs(
+                self.name_starts_with, self.include
+            )
+            for obj in blobs_list:
```

Comment:
Hmm wondering if we would we be able to de-duplicate the code in this method here by doing something like

```
if self.blob:
    blobs_list = [self.blob]
else:
    blobs_list = container_client.list_blobs(...)

for obj in blobs_list:
    ...
```

?